java/ql/src/experimental/semmle/code/java/security/StaticInitializationVectorQuery.qll

```diff
@@ -0,0 +1,166 @@
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.TaintTracking2
+
+/**
+ * Holds if `array` is initialized only with constants, for example,
+ * `new byte[8]` or `new byte[] { 1, 2, 3, 4, 5, 6, 7, 8 }`.
+ */
+private predicate initializedWithConstants(ArrayCreationExpr array) {
+  not exists(array.getInit())
+  or
+  forex(Expr element | element = array.getInit().getAChildExpr() |
+    element instanceof CompileTimeConstantExpr
+  )
+}
+
+/**
+ * An expression that creates a byte array that is initialized with constants.
+ */
+private class StaticByteArrayCreation extends ArrayCreationExpr {
+  StaticByteArrayCreation() {
+    this.getType().(Array).getElementType().(PrimitiveType).getName() = "byte" and
+    initializedWithConstants(this)
+  }
+}
+
+/** Defines a sub-set of expressions that update an array. */
+private class ArrayUpdate extends Expr {
```

Comment:
Should this also cover `java.util.Arrays.fill` and `java.lang.reflect.Array.set...` methods?

Though specifically for this query `Arrays.fill` setting all elements to the same value is probably not what you want as secure IV, so in case you do not add it here it might be good to document the rationale.

And it is unlikely that `Array.set` is used to for creating a random IV, so maybe it is not worth covering that either.

Reply:
I don't think they should be covered. It looks unlikely that these methods are used to set a good IV, although it is technically possible to write code that actually does it. There may be many other methods similar methods. I am not sure what is the best way to document the rationale here. I'll try that. Please let me know in case you have concrete suggestions.

Reply:
The changes look good to me, thanks!